cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,255 @@
+		}
+	}
+	if dups.Len() > 0 {
+		return fmt.Errorf("found tags promoted more than once: %s", dups.List())
+	}
+	return nil
+}
+
+func tagsToDelete(ctx context.Context, client ctrlruntimeclient.Client, promotedTags []api.ImageStreamTagReference, toIgnore sets.String) (sets.String, error) {
+	tagsToCheck := sets.NewString()
+	var errs []error
+	imageStreamsWithPromotedTags := sets.NewString()
+
+	for _, promotedTag := range promotedTags {
+		namespacedImageStream := fmt.Sprintf("%s/%s", promotedTag.Namespace, promotedTag.Name)
+		if imageStreamsWithPromotedTags.Has(namespacedImageStream) {
+			continue
+		}
+		imageStreamsWithPromotedTags.Insert(namespacedImageStream)
+		imagestream := &imagev1.ImageStream{}
+		if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: promotedTag.Namespace, Name: promotedTag.Name}, imagestream); err != nil {
+			if !kerrors.IsNotFound(err) {
+				errs = append(errs, fmt.Errorf("could not get image steam %s in namespace %s", promotedTag.Name, promotedTag.Namespace))
+			}
+			continue
+		}
+		for _, tag := range imagestream.Status.Tags {
+			tagsToCheck.Insert(fmt.Sprintf("%s/%s:%s", promotedTag.Namespace, promotedTag.Name, tag.Tag))
+		}
+	}
+
+	for _, promotedTag := range promotedTags {
+		tagsToCheck.Delete(fmt.Sprintf("%s/%s:%s", promotedTag.Namespace, promotedTag.Name, promotedTag.Tag))
+	}
+	for _, tag := range toIgnore.List() {
+		tagsToCheck.Delete(tag)
+	}
+	if len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+	return tagsToCheck, nil
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	opts := newOpts()
+
+	if err := opts.validate(); err != nil {
+		logrus.WithError(err).Fatal("failed to validate the option")
+	}
+
+	if err := addSchemes(); err != nil {
+		logrus.WithError(err).Fatal("failed to add schemes")
+	}
+
+	abs, err := filepath.Abs(opts.ciOperatorconfigPath)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to determine absolute CI Operator configuration path")
+	}
+	var promotedTags []api.ImageStreamTagReference
+	if err := config.OperateOnCIOperatorConfigDir(abs, func(cfg *api.ReleaseBuildConfiguration, metadata *config.Info) error {
+		promotedTags = append(promotedTags, release.PromotedTags(cfg)...)
+		return nil
+	}); err != nil {
+		logrus.WithField("path", abs).Fatal("failed to operate on CI Operator's config directory")
+	}
+
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", opts.kubeconfig)
+	if err != nil {
+		logrus.WithError(err).Fatalf("could not load kube config from path %s", opts.kubeconfig)
+	}
+
+	client, err := ctrlruntimeclient.New(kubeConfig, ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatalf("could not create client")
+	}
+
+	if err := validatePromotedTags(promotedTags); err != nil {
+		logrus.WithError(err).Fatalf("could not validate promoted tags")
+	}
+
+	toDelete, err := tagsToDelete(context.TODO(), client, promotedTags, opts.ignoredImageStreamTags)
```

Comment:
think about what you would do with `toDelete` when you actually implement deletion - you will need to split it up into name/namespace/tag anyway, so let's use the struct instead of stringifying it

Reply:
Updated.